### PR TITLE
Add cdc chunking support to PebbleCache.

### PR DESCRIPTION
**Pull request by luluz66**
When chunking is enabled, during write, we will first attempt to
chunk it, and then compress each individual chunk, and finally encrypt the chunk
if encryption is enabled.  If the data is compressed, we will first decompress it
before we chunk it.

If there is only one single chunk, we will write the chunk inline,
similar to how we write small files inline when chunking is disabled.
If there are multiple chunks, we will write a file record in Pebble. which include
all the resource names that points to the chunks. Every chunks will be
stored inline in Pebble.

Added unit tests.

This is based on Tyler's PR:
https://github.com/buildbuddy-io/buildbuddy/pull/3457

**Related issues**: N/A

### CODE REVIEW

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 1]
```diff
@@ -1793,6 +1804,196 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
...
+	fileRecord, err := p.makeFileRecord(ctx, r)
+
+	if err != nil {
+		return err
+	}
+	key, err := p.fileStorer.PebbleKey(fileRecord)
+	if err != nil {
+		return err
+	}
+	cdcw.writtenChunks = append(cdcw.writtenChunks, r)
```
Comment (tylerwilliams) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:1961]: maybe worth a comment why this happens first :)

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 2]
```diff
@@ -1793,6 +1804,196 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
...
+	cdcw.eg.Go(func() error {
+		return cdcw.writeRawChunk(fileRecord, key, chunkData)
+	})
+	return nil
+}
+
+func (cdcw *cdcWriter) Write(buf []byte) (int, error) {
+	// If the bytes being written are compressed, we decompress them in
+	// order to generate CDC chunks, then compress those chunks.
+	if cdcw.isCompressed {
```
Comment (tylerwilliams) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:1972]: i wonder if we'll want to do this if the file is gonna be less than 1 chunk size? maybe some way to skip this code path?

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 3]
```diff
@@ -2997,6 +3208,39 @@ type readCloser struct {
 	io.Closer
 }
 
+// newChunkedReader returns a reader to read chunked content.
+// When shouldDecompress is true, the content readed is decompressed.
```
Comment (tylerwilliams) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:3212]:
```suggestion
// When shouldDecompress is true, the content read is decompressed.
```

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 4]
```diff
@@ -1793,6 +1804,199 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
...
+		// file-level metadata entry and a chunk entry into pebble.
+		cdcw.firstChunk = make([]byte, len(chunkData))
+		copy(cdcw.firstChunk, chunkData)
+		return nil
+	}
+
+	if cdcw.numChunks == 2 {
+		if err := cdcw.writeChunkWhenMultiple(cdcw.firstChunk); err != nil {
+			return err
+		}
```
Comment (vadimberezniker) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:1898]: Should we clear out `cdcw.firstChunk` here? I believe the data is no longer needed after this.

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 5]
```diff
@@ -1793,6 +1804,199 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
...
+	if err != nil {
+		return err
+	}
+	key, err := p.fileStorer.PebbleKey(fileRecord)
+	if err != nil {
+		return err
+	}
+
+	// We use cdcw.writtenChunks for the file-level metadata, and this needs to
+	// be in the orther. Otherwise, when we read the file, the chunks will be
```
Comment (vadimberezniker) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:1958]: I think you meant to write "in order" here.

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 6]
```diff
@@ -2744,6 +2959,8 @@ func (e *partitionEvictor) deleteFile(key filestore.PebbleKey, version filestore
 		}
 	case storageMetadata.GetInlineMetadata() != nil:
 		break
+	case storageMetadata.GetChunkedMetadata() != nil:
+		break
```
Comment (vadimberezniker) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:2963]: Do we need to skip updating the partition size in this case?

File: enterprise/server/backends/pebble_cache/pebble_cache.go (hunk 4 above)
Comment (luluz66) [reply enterprise/server/backends/pebble_cache/pebble_cache.go:1902]: Done

File: enterprise/server/backends/pebble_cache/pebble_cache.go (hunk 6 above)
Comment (luluz66) [reply enterprise/server/backends/pebble_cache/pebble_cache.go:2971]: I set the storedSizeBytes in the filemetadata is zero if it's a file level entry that include chunked metadata

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 7]
```diff
@@ -1793,6 +1805,200 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
...
+			return cdcw.writeRawChunk(cdcw.fileRecord, cdcw.key, cdcw.firstChunk)
+		}
+		now := p.clock.Now().UnixMicro()
+
+		md := &rfpb.FileMetadata{
+			FileRecord:      fileRecord,
+			StorageMetadata: cdcw.Metadata(),
+			// The chunks the file record pointed are stored seperately and are
+			// not evicted when this entry is evicted. Therefore, the stored
+			// size bytes should be zero to avoid double counting.
```
Comment (tylerwilliams) [inline enterprise/server/backends/pebble_cache/pebble_cache.go:1876]: these too

**Review by tylerwilliams — APPROVED** [review]

File: enterprise/server/backends/pebble_cache/pebble_cache.go [hunk 8]
```diff
@@ -1793,6 +1805,200 @@ func (p *PebbleCache) Reader(ctx context.Context, r *rspb.ResourceName, uncompre
...
+			return err
+		}
+		if err := cdcw.eg.Wait(); err != nil {
+			return err
+		}
+		if cdcw.numChunks == 1 {
+			// When there is only one single chunk, we want to store the original file
+			// record with the original key instead of computed digest from the
+			// chunkData. This is because the chunkData can be compressed or encrypted,
+			// so the digest computed from it will be different from the original digest.
```
[the remark this answers is not in the document]
Comment (luluz66) [reply enterprise/server/backends/pebble_cache/pebble_cache.go]: Done.

File: enterprise/server/backends/pebble_cache/pebble_cache.go (hunk 7 above)
Comment (luluz66) [reply enterprise/server/backends/pebble_cache/pebble_cache.go:1877]: Done

**Review by vadimberezniker — APPROVED** [review]

**Review by tylerwilliams — APPROVED** [review]

**Pull request merged**